### PR TITLE
Fix #33.  Add user account suspend/reinstate UI.

### LINKED ISSUE #33: Complete or remove the ability to suspend user accounts
There is an attribute of a user, `status`, which can be either `ACTIVE` or `DISABLED`. New users default to `ACTIVE`, there is no UI that I can find that allows the status to be altered, and AFAICT nothing ever reads this value.

Why is it here? Do we still want it? Either fully implement the ability to suspend an account, or remove the dead code and unused database column.

### PR DESCRIPTION
**Pull request by frankduncan**
Issue #33: Complete or remove the ability to suspend user accounts

This is the System UI, as well as login check, only.  There is probably a lot more around account management that we could do, but I believe the spirit of #33 was that we should at least do something with the field in the DB and the related dao functions.  So I think we should hold off on messaging, better error handling, etc until more requirements come, and log issues for those.

One large note is that I found no easy way to force a logout of someone who is suspended while already logged in.  That would involve a DB call on every single page view to check the cms_user table, and add that to the top level spring security interceptors, as the system currently works.  My view was that was overkill for this step of this feature.

Test via logging in as a system user.  Suspend an account.  Ensure you can no longer log in as that account.  Log back in as a system user, reinstate it.  Log back in as other account.

### CODE REVIEW

File: psm-app/cms-web/src/main/java/gov/medicaid/controllers/admin/SystemAdminUserController.java [hunk 1]
```diff
@@ -320,6 +320,43 @@ public ModelAndView delete(@RequestParam("role") String role, @RequestParam("use
...
+     * @throws PortalServiceException - If there are any errors in the action
+     * @endpoint "/system/user/reinstate"
+     */
+    @RequestMapping(value = "/reinstate")
+    public boolean reinstate(@RequestParam("userId") String userId)
+        throws PortalServiceException {
+        CMSUser actor = ControllerHelper.getCurrentUser();
+        getRegistrationService().reinstate(actor, userId);
+
+        return true;
```
Comment (jasonaowen) [inline psm-app/cms-web/src/main/java/gov/medicaid/controllers/admin/SystemAdminUserController.java]: Returning a `boolean` in these functions causes a stack trace:

<details>
  <summary>Unknown return value type: java.lang.Boolean (Click to expand stacktrace)</summary>

```
18:35:53,510 SEVERE [gov.medicaid.controllers.ErrorController] (default task-3) Exception caught by Spring while handling GET request to https://localhost:8443/cms/system/user/reinstate: java.lang.IllegalArgumentException: Unknown return value type: java.lang.Boolean
	at org.springframework.web.method.support.HandlerMethodReturnValueHandlerComposite.handleReturnValue(HandlerMethodReturnValueHandlerComposite.java:79)
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:113)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:827)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:738)
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:85)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:967)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:901)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:970)
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:861)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:846)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at io.undertow.servlet.handlers.ServletHandler.handleRequest(ServletHandler.java:85)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:129)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:317)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:127)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:91)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:114)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:137)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter(AnonymousAuthenticationFilter.java:111)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.authentication.rememberme.RememberMeAuthenticationFilter.doFilter(RememberMeAuthenticationFilter.java:158)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:170)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.savedrequest.RequestCacheAwareFilter.doFilter(RequestCacheAwareFilter.java:63)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:200)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:116)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.csrf.CsrfFilter.doFilterInternal(CsrfFilter.java:100)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.header.HeaderWriterFilter.doFilterInternal(HeaderWriterFilter.java:64)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter.doFilterInternal(WebAsyncManagerIntegrationFilter.java:56)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:105)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:214)
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:177)
	at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:347)
	at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:263)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at org.springframework.web.multipart.support.MultipartFilter.doFilterInternal(MultipartFilter.java:122)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.undertow.servlet.handlers.FilterHandler.handleRequest(FilterHandler.java:84)
	at io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:62)
	at io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)
	at org.wildfly.extension.undertow.security.SecurityContextAssociationHandler.handleRequest(SecurityContextAssociationHandler.java:78)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:131)
	at io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46)
	at io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:64)
	at io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest(AuthenticationMechanismsHandler.java:60)
	at io.undertow.servlet.handlers.security.CachedAuthenticatedSessionHandler.handleRequest(CachedAuthenticatedSessionHandler.java:77)
	at io.undertow.security.handlers.NotificationReceiverHandler.handleRequest(NotificationReceiverHandler.java:50)
	at io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest(AbstractSecurityContextAssociationHandler.java:43)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at org.wildfly.extension.undertow.security.jacc.JACCContextIdHandler.handleRequest(JACCContextIdHandler.java:61)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at org.wildfly.extension.undertow.deployment.GlobalRequestControllerHandler.handleRequest(GlobalRequestControllerHandler.java:68)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:292)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$100(ServletInitialHandler.java:81)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:138)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:135)
	at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48)
	at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)
	at org.wildfly.extension.undertow.security.SecurityContextThreadSetupAction.lambda$create$0(SecurityContextThreadSetupAction.java:105)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1508)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1508)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1508)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1508)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1508)
	at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:272)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$000(ServletInitialHandler.java:81)
	at io.undertow.servlet.handlers.ServletInitialHandler$1.handleRequest(ServletInitialHandler.java:104)
	at io.undertow.server.Connectors.executeRootHandler(Connectors.java:326)
	at io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:812)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```
</details>

File: psm-app/frontend/src/main/js/system/script.js [hunk 2]
```diff
@@ -1325,6 +1325,20 @@ $(document).ready(function () {
         return false;
       });
 
+    $(".suspendLink, .reinstateLink").on('click', function () {
+      var link = $(this);
+      $.ajax({
+        url: link.attr("href"),
+        cache: false,
+        type: "GET",
```
Comment (jasonaowen) [inline psm-app/frontend/src/main/js/system/script.js]: These are modifying application state, so can we make these POST requests? That'll mean handling CSRF; see, for example, [the `postJson` function](https://github.com/SolutionGuidance/psm/blob/3060c7a7bd8b1d97f049879f289ae56242cc812c/psm-app/frontend/src/main/js/script.js#L2144-L2155). I think the relevant headers should be on every page.

File: psm-app/cms-web/src/main/java/gov/medicaid/controllers/admin/SystemAdminUserController.java [hunk 3]
```diff
@@ -320,6 +320,43 @@ public ModelAndView delete(@RequestParam("role") String role, @RequestParam("use
...
+     * @param userId - the user Id
+     * @return the model and view instance that contains the name of view to be rendered and data to be used for
+     *         rendering (not null)
+     *
+     * @throws PortalServiceException - If there are any errors in the action
+     * @endpoint "/system/user/suspend"
+     */
+    @RequestMapping(value = "/suspend")
+    public boolean disable(@RequestParam("userId") String userId)
+        throws PortalServiceException {
```
Comment (jasonaowen) [inline psm-app/cms-web/src/main/java/gov/medicaid/controllers/admin/SystemAdminUserController.java]: Would you mind removing the spurious `throws PortalServiceException` from `RegistrationService#suspend` and `RegistrationService#reinstate`? Then we don't have to have them here, either.

I would probably do it as a separate commit before the functionality commit, and check the rest of the functions in that interface & class.

File: psm-app/cms-business-process/src/main/java/gov/medicaid/services/impl/RegistrationServiceBean.java [hunk 4]
```diff
@@ -821,6 +821,10 @@ public String updateByAdmin(CMSUser actor, CMSUser registrant, String password)
      * @throws PortalServiceException for any errors encountered
      */
     public boolean authenticate(String username, String password) throws PortalServiceException {
+        CMSUser user = findByUsername(username);
+        if (user.getStatus() != UserStatus.ACTIVE) {
+            throw new PortalServiceException("User is disabled");
```
Comment (jasonaowen) [inline psm-app/cms-business-process/src/main/java/gov/medicaid/services/impl/RegistrationServiceBean.java]: Does this message get logged anywhere? Do I have to change some logging configuration to see it?

In the browser, I just see "Invalid username and/or password." -- which, for security reasons, I think is the right message -- but I don't see any error in the server log.

If the answer is just "it always gets swallowed", maybe we should return `false`.

File: psm-app/cms-web/WebContent/WEB-INF/pages/admin/includes/accountsTab.jsp [hunk 5]
```diff
@@ -38,7 +38,12 @@
         </thead>
         <tbody>
             <c:forEach var="item" items="${results.items}">
-            <tr>
+            <tr
+              <c:choose>
+                <c:when test="${item.status == 'DISABLED'}">class="disabledUser"</c:when>
```
Comment (jasonaowen) [inline psm-app/cms-web/WebContent/WEB-INF/pages/admin/includes/accountsTab.jsp]: Probably the logic here should be "If the status is not 'ACTIVE', treat the user as suspended". When there is no status set, both links show up:

```sql
UPDATE cms_user SET status = null WHERE username = 'p2';
```

It's fair to say that's an unusual edge case, but neither the column nor the entity attribute is marked as `not null` (sigh), so we should account for it if we reasonably can.

File: psm-app/frontend/src/main/js/system/script.js [hunk 6]
```diff
@@ -1325,6 +1325,20 @@ $(document).ready(function () {
...
+        dataType: "text",
+        success: function () {
+          link.closest("tr").toggleClass("disabledUser enabledUser");
```
Comment (jasonaowen) [inline psm-app/frontend/src/main/js/system/script.js]: That's super slick!

**Comment by jasonaowen** [comment]
Two things related to this occurred to me today: we should probably send an email notification on suspension/reinstatement, and we should disable "forgot password" resets for suspended accounts. Since the "is suspended?" check happens before the "is correct password?" check, and we don't have a suspension-specific error, the result is fairly confusing for users. Maybe going through the forgot password workflow would result in an email notification to the effect of "Your password was not changed because your account is suspended."

Those could probably happen later, but if so we should avoid closing the issue. (Or, open follow-up issues.)

File: psm-app/cms-web/src/main/java/gov/medicaid/controllers/admin/SystemAdminUserController.java (hunk 3 above)
Comment (frankduncan) [reply psm-app/cms-web/src/main/java/gov/medicaid/controllers/admin/SystemAdminUserController.java]: Definitely.  I didn't clean up anything else in that class because after looking at it, it proved to be a bit too deep of a rabbit hole and should be saved for whitespace cleanups later.

Comment (frankduncan) [reply psm-app/cms-web/src/main/java/gov/medicaid/controllers/admin/SystemAdminUserController.java]: We should keep in our minds to do this across the codebase.

I understand ensuring interfaces giving implementations openings to throw exceptions, but this feels like a bit too much.

File: psm-app/cms-business-process/src/main/java/gov/medicaid/services/impl/RegistrationServiceBean.java (hunk 4 above)
Comment (frankduncan) [reply psm-app/cms-business-process/src/main/java/gov/medicaid/services/impl/RegistrationServiceBean.java]: I definitely punted this to the "architectural question."  I think it's most correct to call this an exceptional situation, since I think authenticate should return true if user/password is correct, false if user/password is incorrect, and use the exception framework to handle other cases, to give full information to the caller about what exactly is the nature of the failure.  

\<tangent>Indeed, I think there should be a different exception in the case that the user isn't in the database, since I place giving information to the user that they are using an incorrect login name as a usability win trumps security concerns about some malicious person trying to figure out what usernames are registered. \</tangent>

But realistically the caller does nothing with this information, so simplifying the code to be in line with actual use is reasonable as well.

Your call :)

File: psm-app/cms-web/WebContent/WEB-INF/pages/admin/includes/accountsTab.jsp (hunk 5 above)
Comment (frankduncan) [reply psm-app/cms-web/WebContent/WEB-INF/pages/admin/includes/accountsTab.jsp]: Better option to add a migration to make it not null?  That would be my preference.

**Comment by frankduncan** [comment]
> we should probably send an email notification on suspension/reinstatement

I can open an issue, since there's some design discussion here.  I can make a cogent case for either side of whether users should be alerted that they've been suspended/reinstated.

> we should disable "forgot password" resets for suspended accounts

This depends on whether we want a different workflow for suspended accounts, and whether we want that information to be presented to the user.  If you think so, I'll open an issue for it.

File: psm-app/cms-web/src/main/java/gov/medicaid/controllers/admin/SystemAdminUserController.java (hunk 1 above)
Comment (frankduncan) [reply psm-app/cms-web/src/main/java/gov/medicaid/controllers/admin/SystemAdminUserController.java]: Ack!  Thanks for the catch!

**Comment by frankduncan** [comment]
Thanks @jasonaowen !  Ready for re-review and discussion :)

File: psm-app/cms-web/src/main/java/gov/medicaid/controllers/admin/SystemAdminUserController.java (hunk 3 above)
Comment (jasonaowen) [reply psm-app/cms-web/src/main/java/gov/medicaid/controllers/admin/SystemAdminUserController.java]: Yeah, I just did something in #959, and it did end up being a bit bigger than I initially expected. You're right that we should do a similar pass for `RegistrationService`, but I don't think we need to do the whole thing as part of this PR - just these two methods. Of course, if you want to do more, that'd be great, too! :D

File: psm-app/cms-web/WebContent/WEB-INF/pages/admin/includes/accountsTab.jsp (hunk 5 above)
Comment (jasonaowen) [reply psm-app/cms-web/WebContent/WEB-INF/pages/admin/includes/accountsTab.jsp]: Sure, that'd be even better!

File: psm-app/frontend/src/main/js/system/script.js (hunk 2 above)
Comment (jasonaowen) [reply psm-app/frontend/src/main/js/system/script.js]: Oof, thank you! Now that you mention it, I think we do that in a few other places, too; I should go open some issues.

File: psm-app/cms-business-process/src/main/java/gov/medicaid/services/impl/RegistrationServiceBean.java (hunk 4 above)
Comment (jasonaowen) [reply psm-app/cms-business-process/src/main/java/gov/medicaid/services/impl/RegistrationServiceBean.java]: >  I think it's most correct to call this an exceptional situation, since I think authenticate should return true if user/password is correct, false if user/password is incorrect, and use the exception framework to handle other cases, to give full information to the caller about what exactly is the nature of the failure.

That's reasonable! Mostly I was curious how it works; I didn't dig into it at all, and wondered if you had.

If we were designing this from a clean slate, I think I would want to have suspended users be able to successfully log in but only see a "Account is suspended" screen, maybe with a link to change the password. But, we're not, and that's more work than this feature is worth right now.

I don't think this needs to change! Thank you for sharing your thinking. :)

**Comment by jasonaowen** [comment]
> > we should probably send an email notification on suspension/reinstatement
> 
> I can open an issue, since there's some design discussion here.  I can make a cogent case for either side of whether users should be alerted that they've been suspended/reinstated.
> 
> > we should disable "forgot password" resets for suspended accounts
> 
> This depends on whether we want a different workflow for suspended accounts, and whether we want that information to be presented to the user.  If you think so, I'll open an issue for it.

You're right, those are both design questions, not implementation questions! Would you mind opening (an) issue(s)?
